nxcomp/Log.h

```diff
@@ -0,0 +1,526 @@
+	size_t line() const
+	{
+		return line_;
+	}
+
+	/** Severity level of the event */
+	NXLogLevel level() const
+	{
+		return level_;
+	}
+
+	/** Time of the event */
+	struct timeval timestamp() const
+	{
+		return timestamp_;
+	}
+
+
+	NXLogStamp(const char *file, const char *function, int line, NXLogLevel level)
+	{
+		file_      = std::string(file);
+		function_  = std::string(function);
+		line_      = line;
+		level_     = level;
+		gettimeofday(&timestamp_, NULL);
+	}
+
+};
+
+
+/**
+ * Log class
+ *
+ * Logs events to a stream, filters by file/level
+ */
+class NXLog
+{
+	NXLogLevel level_;
+
+	std::ostream *stream_;
+	std::map< std::string, NXLogLevel > per_file_levels_;
+	bool synchronized_;
+	size_t  thread_buffer_size_;
+	pthread_mutex_t output_lock_;
+	pthread_key_t tls_key_;
+	NXLogLevel breakpoint_level_;
+	bool breakpoint_enabled_;
+
+	bool log_level_;
+	bool log_time_;
+	bool log_unix_time_;
+	bool log_location_;
+	bool log_thread_id_;
+
+	typedef struct per_thread_data_s
+	{
+		NXLogLevel          current_level;
+		std::string*        current_file;
+		std::string*        thread_name;
+		std::stringstream*  buffer;
+		NXLog*              log_obj;
+	} per_thread_data;
+
+
+	static void free_thread_data(void* arg)
+	{
+		per_thread_data *pdt = (per_thread_data*)arg;
+
+		if ( !pdt )
+			return;
+
+		if ( pdt->log_obj ) {
+			// Ensure the buffer is flushed before thread deletion
+			pdt->log_obj->flush(pdt);
+		}
+
+		delete pdt->current_file;
+		delete pdt->thread_name;
+		delete pdt->buffer;
+		delete pdt;
+	}
+
+	per_thread_data* get_data_int() const
+	{
+		per_thread_data *ret = NULL;
+ 
+		if ( (ret = (per_thread_data*)pthread_getspecific(tls_key_)) == NULL )
+		{
+			ret = new per_thread_data;
+			ret->current_level = NXDEBUG;
+			ret->buffer = new std::stringstream();
+			ret->current_file = new std::string();
+			ret->thread_name = new std::string();
+			ret->log_obj = const_cast<NXLog*>(this);
+			pthread_setspecific(tls_key_, ret);
+		}
+
+		return ret;
+	}
+
+	per_thread_data* get_data()
+	{
+		return get_data_int();
+	}
+
+	const per_thread_data* get_data() const
+	{
+		return get_data_int();
+	}
+
+	/** Convert NXLogStamp to string according to the current configuration */
+	std::string stamp_to_string(const NXLogStamp& stamp) const;
+
+	/**
+	 * Internal flush function
+	 *
+	 * When a thread is being terminated and free_thread_data gets called,
+	 * the TLS key gets set to NULL before the call to free_thread_data,
+	 * and the destructor function gets the old value. 
+	 *
+	 * This means that get_data() stops working correctly, and we need
+	 * to be able to pass the old pointer.
+	 */
+	void flush(per_thread_data *pdt)
+	{
+		const std::string str = pdt->buffer->str();
+
+		if (!str.empty())
+		{
+			pthread_mutex_lock(&output_lock_);
+			(*stream()) << str;
+			pthread_mutex_unlock(&output_lock_);
+			pdt->buffer->str("");
+
+			if ( breakpoint_enabled_ && pdt->current_level <= breakpoint_level_)
```

Comment:
That's a hack that will only work on `x86`/`x86_64` hardware. I hope it's scheduled to be removed (together with `breakpoint_level()` and friends) before the actual merge?